### ISSUE TITLE
Manipulating Objects

### ISSUE THREAD

**Issue by mike-thompson-day8** [rethinkdb/rethinkdb#5707]
**First**, this is a docs request ... 

Based on my recent experience coming up to speed on manipulating objects, I'd suggest that the following example should be included somewhere in the API docs.  Had I seen it ... would have saved me a lot of time.  

```
r({"a": 1, "b": 2}).do(
  (obj) =>  r.map(obj.keys(), obj.values(), (k, v) => [k, v]).coerceTo('OBJECT'))
```

For me, if this example had been in the docs for `map`, it would have been gold. Because that's where I first went looking, trying to understand how I could iterate across all key/value pairs. 

**Second** this is a new function suggestion ... 

Although the above works, I feel like it requires a lot of boilerplate. it is noisy.  I'm relatively new to rethinkdb, so perhaps this business of iterating across all key/values pairs doesn't have to happen very often, so I may be out of line, but I would have thought this was a sufficiently common case for the existence of special function `mapo`  which could be like this (equivalent to the code above):

```
r({"a": 1, "b": 2}).do(
  (obj) =>  obj.mapo((k, v) => [k, v]))
```

All the boilerplate around `keys` and `values` and `coerceTo` would be removed.   

Perhaps the better name should be `map-kv` 

Just a thought.

**Comment by danielmewes** [comment]
Thanks for the proposal @mike-thompson-day8 . Having a special command for this makes sense I think.

My suggestion for a name is `map_object`.

One question is whether the function should be able to manipulate keys or just values. In the simplest case, the mapping function `function(k, v) -> v'` would return the new value for the key `k`, but couldn't change the key itself. This avoids a bunch of edge cases, but is obviously less powerful.

As a next step, we can allow the function to return a pair `[k', v']` like in your proposal that might rename the key, but cannot delete it. Finally, we could also allow returning a `null` key such as `[null, v]` to drop a key from the result.

We'll try to discuss this proposal for one of the upcoming releases.

**Comment by marshall007** [comment]
I think this term should be named something like `map_values`. I think the name `map_object` might confuse people into thinking `.map(...)` doesn't work on objects. I also think the signature should be three-arguments: `function(val, key, row)` with `val` as the first argument since that's the only argument you need in the most common case.

@danielmewes: I'm not sure I'm a fan of returning `[k', v']` as a way to rename/drop a key. You would run into problems if you were trying to build array values or if your value is already an array. I certainly like the idea and it would be really useful, but I think the behavior is a unexpected. Perhaps this could be prevented by wrapping the value in an `r.literal(...)`, but I still think that's a bit subtle.

**Comment by mlucy** [comment]
@danielmewes -- there's an alternate proposal at https://github.com/rethinkdb/rethinkdb/issues/878 for just overloading things more.

**Comment by danielmewes** [comment]
I'm not a huge fan of #878 because the expected behavior described there clashes with the one described in https://github.com/rethinkdb/rethinkdb/issues/2768 . So if we ended up overloading `map` for objects it would likely work correctly for ~half of the users, and would work in a really confusing way for the other half who expect `map` on a single selection to work like `do`.

**Comment by mlucy** [comment]
I think #2768 is probably a bad idea.  You don't want to make `map` act like `do`, because then you have two terms which are 99% the same, except that one treats arrays as a single datum and the other treats arrays as a sequence.  (Worse, this is a subtle distinction that isn't at all obvious from the names.)

It's also extremely unusual: no other language that I know of makes `map` turn into a function call when called on a non-sequence.

By contrast, I think it's actually fairly common for functional languages to define `map` on objects.  In Ruby, for example:
```
irb(main):076:0> {a: 1}.map{|k,v| [k,v]}
[[:a, 1]]
```

Or in JQuery:
```
var dimensions = { width: 10, height: 15, length: 20 };
var keys = $.map(dimensions, function(val, key) {
  return [key, val];
});
```

---

Personally, I would expect calling `map` on an object to just work (although it's a little bit debatable whether it should return an array or another object).

Also, we've in general tried to add new functionality to ReQL by adding new behavior to existing commands rather than by introducing new commands for one special use case.  This has made the language a little more complicated to implement, but I think in the long run it's been a very good thing, because it's stopped the number of commands from ballooning out of control every time we get a new use case.

When there's an opportunity to satisfy a specific use case by adding a new general rule that makes sense (like #878, which defines a general rule for applying stream operations like `map` or `filter` to objects), I think that's usually the right way to do it.

**Comment by danielmewes** [comment]
I agree that #2768 is probably a bad idea, but it shows that some users expect `map` to work on a single selection the same way it does on a stream.

I even think that this expectation is justified. If I see two queries `r.table("t").get("a").map(...)` and `r.table("t").getAll("a").map(...)`, I wouldn't expect the `map` in there to do different things.

I think overloads are great if their meaning is obvious. I think they are not great if different people have different expectations as to what the "natural" semantics are. I think mapping over an object and mapping over a sequence are sufficiently different operations that they should have different terms. 

Even the signature of the commands is different. For a sequence, the mapping function receives one argument. For an object, it receives two arguments (key and value).

Overloading `map` makes sense in many languages. But I think it's not a good idea in ReQL. If we didn't have a `SINGLE_SELECTION`, I would probably think differently about this. But since we do, and the difference between `get` and `getAll` keeps confusing new users even as it is, my opinion is that we shouldn't do the overload.

My opinion is that we should do neither #2768 nor #878 because of this ambivalence.

**Comment by marshall007** [comment]
@danielmewes @mlucy I think #878 would make a lot of sense if we started down the path of complete `SEQUENCE <-> SINGLE_SELECTION` polymorphism. In doing so I think the `map` case would be more obvious since you would expect all terms that operate on sequences to behave a certain way for objects. Some notable examples:

```js
.count(function(v, k))
.distinct(function(v, k))
.reduce(function(acc, v, k))
.map(function(v, k))
```

> Personally, I would expect calling `map` on an object to just work (although it's a little bit debatable whether it should return an array or another object).

I think `object.map(...)` should return an array and we should still implement a new term like the one proposed here for returning an object.

**Comment by mlucy** [comment]
> I think object.map(...) should return an array and we should still implement a new term like the one proposed here for returning an object.

That's an interesting idea; that's what Ruby and JQuery appear to be doing.

**Comment by mlucy** [comment]
> I agree that #2768 is probably a bad idea, but it shows that some users expect map to work on a single selection the same way it does on a stream.

That's fair.  On the other hand, I think it's inevitable in any language with overloading that people will sometimes misunderstand what the overloading does.  I might be biased because I use a language where `map` is defined on objects, but it honestly seems like the clearest possible interpretation of `object.map(func)` is mapping over the object, not calling `func(object)`.
